language-plutus-core/generators/Language/PlutusCore/Generators/AST.hs

```diff
@@ -1,99 +1,161 @@
-          errorGen = Error () <$> genType
-          recursive = [absGen, instGen, lamGen, applyGen, unwrapGen, wrapGen]
-          nonRecursive = [varGen, Constant () <$> genConstant, Builtin () <$> genBuiltin, errorGen]
-
-genProgram :: MonadGen m => m (Program TyName Name ())
+genConstant :: AstGen (Constant ())
+genConstant = Gen.choice [genInt, genBS] where
+    int' = Gen.integral_ (Range.linear (-10000000) 10000000)
+    string' = BSL.fromStrict <$> Gen.utf8 (Range.linear 0 40) Gen.unicode
+    genInt = makeBuiltinInt <$> int'
+    genBS = BuiltinBS () <$> string'
+
+genType :: AstGen (Type TyName ())
+genType = simpleRecursive nonRecursive recursive where
+    varGen = TyVar () <$> genTyName
+    funGen = TyFun () <$> genType <*> genType
+    lamGen = TyLam () <$> genTyName <*> genKind <*> genType
+    forallGen = TyForall () <$> genTyName <*> genKind <*> genType
+    applyGen = TyApp () <$> genType <*> genType
+    recursive = [funGen, applyGen]
+    nonRecursive = [varGen, lamGen, forallGen]
+
+genTerm :: AstGen (Term TyName Name ())
+genTerm = simpleRecursive nonRecursive recursive where
+    varGen = Var () <$> genName
+    absGen = TyAbs () <$> genTyName <*> genKind <*> genTerm
+    instGen = TyInst () <$> genTerm <*> genType
+    lamGen = LamAbs () <$> genName <*> genType <*> genTerm
+    applyGen = Apply () <$> genTerm <*> genTerm
+    unwrapGen = Unwrap () <$> genTerm
+    wrapGen = IWrap () <$> genType <*> genType <*> genTerm
+    errorGen = Error () <$> genType
+    recursive = [absGen, instGen, lamGen, applyGen, unwrapGen, wrapGen]
+    nonRecursive = [varGen, Constant () <$> genConstant, Builtin () <$> genBuiltin, errorGen]
+
+genProgram :: AstGen (Program TyName Name ())
 genProgram = Program () <$> genVersion <*> genTerm
+
+{- Note [Name mangling]
+We want to test that turning a term into a distinct one results in a failed equality check.
+For this we keep the spine of the term the same, but change some of its variables at their
+usage sites. Variables that are going to be changed are selected before the mangling happens,
+so that this subset of term's variables can be easily controlled and is dependent on the size
+parameter of the generator. Once variables are selected, the next step is to generate some new
+variables none of which is a member of the set of variables prepared for mangling (but the new
+variables are allowed to overlap with those that the term already contains and that are not
+going to be mangled). The last step is to actually mangle the term by replacing /each usage
+of a variable/ from the prepared set of variables with a /random/ variable from the set of new
+variables. This way we get diverse and interesting mangled terms.
+-}
+
+subset1 :: (MonadGen m, Ord a) => Set a -> m (Maybe (Set a))
+subset1 xs
+    | size == 0 = return Nothing
+    | otherwise = fmap Just . Gen.set (Range.linear 1 size) . Gen.element $ Set.toList xs
+    where size = Set.size xs
+
+substAllNames
```

Comment:
This seems similar to the functions in `Substitute`, just monadic. Can we add it there and/or make them monadic? Seems like a reasonable generalization and probably useful in other places.

Reply:
I'd keep it in this file for now. It's of this form:

```haskell
termSubstNamesM
    :: (Name () -> Maybe (m (Name ())))
    -> Term TyName Name ()
    -> m (Term TyName Name ())
```

but I'd expect a building block to be of this form:

```haskell
termSubstNamesM
    :: (Name () -> Maybe (m (Name ())))
    -> Term TyName Name ()
    -> m (Term TyName Name ())
```

and the latter in this particular case would mean that we need to generate a new name for each variable and only then decide whether we actually want to rename that variable, doesn't sound good.

Reply:
Hm, no, I think I'm wrong.

Reply:
Yeah, you're right. But it's taking more time than I expected, I'll finish tomorrow.

Reply:
Done.